{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7679d56e-a626-486c-b42f-8f93bfc06d9a",
   "metadata": {},
   "source": [
    "# Extra Material and Practice: Functions, NumPy, Pandas, and Matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee739f2-06d1-4d33-989b-dc8be454f366",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import DS libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sn\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca13334d-af4f-4ec4-b67f-ce81dc5e0bdb",
   "metadata": {},
   "source": [
    "## Data sets for Extra Material\n",
    "1. malariaMaize.csv\n",
    "\n",
    "2. mammals.csv\n",
    "    * find it here: https://zief0002.github.io/bespectacled-antelope/codebooks/mammals.html\n",
    "    * this notebook will demo of how to read it into your notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de5fe41a-38c2-4440-b7ae-1379212d638e",
   "metadata": {},
   "source": [
    "# Functions: Extra Material\n",
    "---\n",
    "\n",
    "## Extra Examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dd1e8a5-2a2b-4587-8f18-063005319055",
   "metadata": {},
   "outputs": [],
   "source": [
    "# I stole this example from code academy\n",
    "# You may recognize this example from a previous extra material notebook - now we've made it into a function\n",
    "# this example explicitly points out some simple ways to use if/elif/else statements\n",
    "# with user-inputted information to ensure that the user submitted what was expected!\n",
    "# this function does not have a explicitly defined return statement. It still is returning something -- it is printing to the screen \n",
    "\n",
    "def clinic():\n",
    "    print(\"You've just entered the clinic!\")\n",
    "    print(\"Do you take the door on the left or the right?\")\n",
    "    answer = input(\"Type left or right and hit 'Enter'.\").lower()\n",
    "    if answer == \"left\" or answer == \"l\":\n",
    "        print(\"This is the Verbal Abuse Room, you heap of parrot droppings!\")\n",
    "    elif answer == \"right\" or answer == \"r\":\n",
    "        print(\"Of course this is the Argument Room, I've told you that already!\")\n",
    "    else:\n",
    "        print(\"You didn't pick left or right! Try again.\")\n",
    "              \n",
    "# Call the clinic function here!\n",
    "# remember that there is a difference between defining a function - as we did above -\n",
    "# and ACTUALLY CALLING IT.\n",
    "clinic()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e88af93-8bb8-4edc-8d0f-4b9dc15ae4b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3628800\n"
     ]
    }
   ],
   "source": [
    "#an example of counting down with a while loop\n",
    "def factorial(x):\n",
    "    fac=1\n",
    "    while x!=0:\n",
    "        fac=fac*x\n",
    "        x=x-1\n",
    "    return fac\n",
    "print(factorial(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3949f858-ae81-4836-9d35-3c4621487e81",
   "metadata": {},
   "source": [
    "[**The Monty Hall Problem**](https://rosettacode.org/wiki/Monty_Hall_problem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2307a360-c12b-492c-8fe4-0055bee6cdc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The Monty Hall problem. I stole this from Rosetta code which is a website that translates popular\n",
    "# programming problems into multiple languages: \n",
    "# https://rosettacode.org/wiki/Monty_Hall_problem#Python_3_version:\n",
    "# we haven't talked about modules and libraries yet, but you can think of them as externally bundled functions.\n",
    "# Externally bundling functions with other related functions allows them to be encapsulated and reused. \n",
    "import random\n",
    " #1 represents a car\n",
    " #0 represent a goat\n",
    " \n",
    "stay = 0  #amount won if stay in the same position\n",
    "switch = 0 # amount won if you switch \n",
    " \n",
    "for i in range(1000):\n",
    "    lst = [1,0,0]           # one car and two goats\n",
    "    random.shuffle(lst)     # shuffles the list randomly\n",
    " \n",
    "    ran = random.randrange(3) # gets a random number for the random guess\n",
    " \n",
    "    user = lst[ran] #storing the random guess \n",
    " \n",
    "    del(lst[ran]) # deleting the random guess\n",
    " \n",
    "    huh = 0\n",
    "    for i in lst: # getting a value 0 and deleting it\n",
    "        if i ==0:\n",
    "            del(lst[huh]) # deletes a goat when it finds it\n",
    "            break\n",
    "        huh+=1\n",
    " \n",
    "    if user ==1: # if the original choice is 1 then stay adds 1\n",
    "        stay+=1\n",
    " \n",
    "    if lst[0] == 1: # if the switched value is 1 then switch adds 1\n",
    "        switch+=1\n",
    " \n",
    "print(\"Stay =\",stay)\n",
    "print(\"Switch = \",switch)\n",
    "#Done by Sam Witton 09/04/2014\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8ee0f56-b9b0-4579-a0b8-222ae8904e6b",
   "metadata": {},
   "source": [
    "### Functions can change the values of lists\n",
    "Functions usually don't change the values of passed parameters however, if you are using lists - which contain pointers rather than actual values, functions that are called CAN change the value of elements of the list. \n",
    "\n",
    "Let's look at a comparison:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6045c847-f425-451b-95ee-a95bfd8205c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here the input is not a list\n",
    "def change_num_by_two(num):\n",
    "    #num = 2*num\n",
    "    num *= 2\n",
    "    print(\"Inside the function. The value of num is: \")\n",
    "    print(num)\n",
    "    return num\n",
    "\n",
    "mynum = 7\n",
    "mynum2 = change_num_by_two(mynum)\n",
    "\n",
    "print(\"after being sent through the change_by_two function, the value of mynum2 set in the program is: \")\n",
    "print(mynum2)\n",
    "print(\"but the original value of mynum set in the program is still: \")\n",
    "print(mynum)\n",
    "print(\"--------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f10df6eb-5ce4-4a23-9e84-12f3cefe4a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compare this to passing a list\n",
    "def change(my_list):\n",
    "    if(len(my_list)>0):\n",
    "        print(\"did I get here?\")\n",
    "        my_list[0]=my_list[0]*2\n",
    "        print(my_list[0])\n",
    "    else:\n",
    "        print(\"there is nothing in this list\")\n",
    "        \n",
    "mylist=[7,6,5,4,3]\n",
    "empty_list=[]\n",
    "\n",
    "print(\"Here is the first element of the list before we call the function: \")\n",
    "print(mylist[0])\n",
    "\n",
    "# Calling the function change on two different lists:\n",
    "change(mylist)\n",
    "change(empty_list)\n",
    "\n",
    "#------------\n",
    "\n",
    "print(\"Here is the first element of the list after we call the function: \")\n",
    "print(mylist[0])\n",
    "print(mylist)\n",
    "# Has the list been changed outside of the function? Is that what you would expect? Why or why not? We can revist pythontutor.org for \n",
    "# a hint on why lists can be modified from within a loop or a function... whereas other variables cannot be. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d22e6af-46f0-4079-be41-ab3133bc0df4",
   "metadata": {},
   "source": [
    "### Functions can call *other* functions\n",
    "In fact, Functions can call *THEMSELVES*. This is called recursion, and it is used to solve particular problems, like in phylogeny construction. \n",
    "\n",
    "I really want you to work through these examples and make sure you understand where the program is calling/running in the code. PythonTutor will probably help with this visualization. \n",
    "\n",
    "#### Example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dcfdb2c-faad-46e1-8f82-5ff49fd97592",
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_good_turn(n):\n",
    "    return n + 1   \n",
    "\n",
    "def deserves_another(n):\n",
    "    #return one_good_turn(n+2)\n",
    "    return one_good_turn(n) + 2\n",
    "\n",
    "#main body of the program    \n",
    "print(deserves_another(10)) #calls the function deserves_another"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5964541-9614-44d7-856a-8dcb8bf09690",
   "metadata": {},
   "source": [
    "## Extra Material: Revisiting Scope \n",
    "\n",
    "* Scope  - is the region of the program from which a variable can be referenced. It is the hierarchical order in which namespace is searched to find the object.\n",
    "  * **L.E.G.B** is the acronym for how scope officially works: Local, Enclosed, Global and Built-in.\n",
    "    * **Local** can only be 'reached' by their local environment (Function)\n",
    "    * **Enclosed** is accessed from inside nested function\n",
    "    * **Global** is accessed anywhere in the program\n",
    "    * **Built-in** are reserved names in Python (like keywords such as 'def').\n",
    "<br><br>\n",
    "\n",
    "* The only way that a function can see a variable from another function (including the main body of the program itself) *is if the variable is passed as an argument/parameter to the function itself*\n",
    "\n",
    "* Local variables (within the function) cease to exist once the function is completed - the local variables are released from memory once the function is finished (so the memory is erased).\n",
    "\n",
    "* **Potential Confusion Warning:** We saw the idea of scope when we discussed loops. Keep in mind that variables that are passed into a function don’t change their value outside of the function (i.e. in the main body of the program)…usually. This is because most variables that we are interested in using are immutable such as numbers and tuples (python passes them ‘by value’). However, we were also talking about *iterating over lists* and confusingly, lists work a little bit differently than regular variables because they are made up of pointers rather than actual data/variables. Thus, they can be manipulated/changed from within a loop (called 'changed in place') and maintain that difference even once the loop is finished. Mutable data types such as lists CAN HAVE THEIR VALUE CHANGED by a function and passed back to the call where **it will remain changed.** However, this is not true of most variables since most variable are immutable."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f129ee04-ed29-4e13-b0d5-089095a63a42",
   "metadata": {},
   "source": [
    "### Example\n",
    "An example illustrating a simple function wherein I sneak in some Bayes' theorem because I love Bayes' and also because it harbours a fundamentally important philophy to computing, being able to update a prior probability/expectation based on new information.\n",
    "\n",
    "--------------------------------------\n",
    "**BAYES IS SUPER IMPORTANT IN COMPUTING (especially with biological data)**\n",
    "\n",
    "----------------------------------------\n",
    "Functionally: it allows you to update your probability when you encounter new information or new evidence.\n",
    "\n",
    "Often we see Bayes' in the context of [sensitivity and specificity](https://en.wikipedia.org/wiki/Sensitivity_and_specificity) of tests, like the tests for a certain disease.\n",
    "\n",
    "#### Disease Test Example: Scenario\n",
    "\n",
    "Let's imagine that you are looking for an early indicator of a type of cancer. This cancer, call it cancer \"A\", is found at 1.5% of a population. Your research group has found a biomarker that is predictive of 80% of sufferers of this cancer.\n",
    "\n",
    "Most people (and, yes, sadly this does include physicians, who have been surveyed about this repeatedly and still mostly get it wrong) would conclude: \"80%?! That's great. That means if there is a positive test for this biomarker, the individual has an 80% chance of having cancer \"A\"\".\n",
    "\n",
    "However, THAT IS DEFINITELY NOT the correct interpretation.\n",
    "\n",
    "Is this surprising to you?\n",
    "\n",
    "In order to get to the correct interpretation, we need to know the answer to another question:\n",
    "\n",
    "**How many individual without the disease would get a positive result for this biomarker?**\n",
    "\n",
    "#### Disease Test Example: Analysis\n",
    "Let's plug in some numbers to illustrate the problem:\n",
    "\n",
    "---\n",
    "\n",
    "False positive rate = 4% <br>\n",
    "True negative rate = 96%\n",
    "\n",
    "---\n",
    "\n",
    "True positive rate = 80% <br>\n",
    "False negative rate = 20%\n",
    "\n",
    "---\n",
    "\n",
    "So, this test will only fail to identify 20 out of 100 sufferers, and it will only misdiagnose 4 out of 100 healthy individuals. Still sounds good, eh?\n",
    "\n",
    "Ah, but no.\n",
    "\n",
    "Let's use round numbers: We have a population of 10,000 individuals. We expect 1.5% of them to suffer from cancer A which means about 150 individuals have cancer A. Of this 150 sufferers, 120 of them will - accurately - get a positive test for the biomarker (based on the 80% true positive rate).\n",
    "\n",
    "But 9850 people in this population don't have cancer \"A\". And, with just a 4% false positive rate (which is quite low), we still expect there be 394 to test positive!\n",
    "\n",
    "If you are patient who has recieved a positive result, you will want to know \"what is the chance that i actually have cancer A?\".\n",
    "\n",
    "You have to **update** your probability. You walked into the clinic with a 1.5% chance of having cancer A, but now - with a positive test result - you have updated your probability to approximately 23%. This percentage comes from the fact that there are 514 positive individuals over all, (120 true positives and 394 false positives).  \n",
    "\n",
    "Notably, this is not 80%, though!\n",
    "\n",
    "(*Note:* these numbers are for the CA-125 biomarker used to detect ovarian cancer, with a lifetime risk for women of developing ovarian cancer is 1.34% and it has a false positive rate of 4%)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e12e1208-1f5f-4379-9c43-d1f5dbea470d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def biomarker(pDisease, pPosDisease, pPosNoDisease):\n",
    "    pNoDisease =1.0-pDisease\n",
    "    pPos=pPosDisease*pDisease+pPosNoDisease*pNoDisease\n",
    "    return round((pPosDisease*pDisease)/pPos,3)\n",
    "\n",
    "pDisease=0.015\n",
    "pPosDisease=0.8\n",
    "pPosNoDisease=0.04\n",
    "print(\"Probability(disease|positive result)= \", biomarker(pDisease, pPosDisease, pPosNoDisease))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43a8b269-f2cd-4b75-8585-76c13cb9ffa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A simple and generalizable Bayesian function with Python String format\n",
    "def bayes(outComeA,outComeB,pB,pAGivenB,pAGivenNotB):\n",
    "    pNotB=1.0-pB\n",
    "    pA=pAGivenB*pB+pAGivenNotB*pNotB\n",
    "    pBGivenA=(pAGivenB*pB)/pA\n",
    "    return 'p (%s|%s) = %.2f' %(outComeB,outComeA,pBGivenA)\n",
    "\n",
    "#main part of program\n",
    "geneName='TP53 Tumor protein p53 [Homo sapiens]'\n",
    "geneID=7157\n",
    "matchProbability=98.64756341\n",
    "#print(\"The gene to be analyzed is: %s\" % geneName)\n",
    "print(\"The gene number is: %d: %s\"%(geneID,geneName))\n",
    "\n",
    "outComeA=\"Positive test result\"\n",
    "outComeB=\"has disease\"\n",
    "pB=0.015\n",
    "pAGivenB=0.8\n",
    "pAGivenNotB=0.04\n",
    "print(bayes(outComeA,outComeB,pB,pAGivenB,pAGivenNotB))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd741f90-bfa0-4e86-b984-9c6b23d473f8",
   "metadata": {},
   "source": [
    "## Functions: Extra Practice Questions\n",
    "\n",
    "1. PSEUDOCODE: A DNA string is called an open reading frame (ORF) if it begins with 'ATG', ends with 'TGA', 'TAG', or 'TAA', and has a length that is a multiple of 3. ORFs are interesting because they can encode proteins.\n",
    "\n",
    "    Write a function called `ORFadviser(DNA)` that takes a string called DNA as input and works as follows:\n",
    "  \n",
    "    1. The function returns the string 'This is an ORF.' if the input string satisfies all three of the conditions required of ORFs.\n",
    "    2. Otherwise, if the first three symbols are not 'ATG', the function returns the string 'The first three bases are not ATG.'.\n",
    "    3. Otherwise, if the string does not end with 'TGA', 'TAG', or 'TAA', the function returns the string 'The last three bases are not a stop codon.'.\n",
    "    4. Otherwise, the function returns the string 'The string is not of the correct length.' <br><br>\n",
    "\n",
    "2. PSEUDOCODE: Let's assume that we have a long DNA sequence in which each symbol is random and equally likely to be an A, C, T, or G. We'd like to determine the average number of symbols until we see the first occurrence of a start codon \"ATG\".  We'll break this into parts:\n",
    "\n",
    "    1. First, write a function in pseudocode called `lengthToATG()` that takes no input, generates a random letter from \"A\", \"T\", \"C\", or \"G\", and then if the last three letters generated are \"ATG\", it returns the total number of symbols that were generated. Otherwise, it repeats the process by generating a next random letter.\n",
    "\n",
    "    2. Compute the average time to the ATG by creating another function that takes `lengthToATG()` as an argument and runs it *n* number of trials to see when an ATG is created (n=10 or 100 or 1000).\n",
    "\n",
    "3. Imagine that we have a list of DNA sequences and we wish to know how many of them are of a specific length. Write a function called `countLength(DNAlist, length)` that takes a list of DNA strings called `DNAlist` and a positive integer `length` as input and returns the number of strings in the list that have the specified length. Use the built-in `len()` function and for loops. Here's an example of the function in action:\n",
    "\n",
    "        countLength([\"ATA\", \"ATCG\", \"TTT\", \"A\"], 3)  \n",
    "        return: 2 (because there are two sequences, ATA and TTT, that are 3 nucleotides long)\n",
    "<br>\n",
    "\n",
    "4. Next, write a function called `getLength(DNAlist, length)` that is analogous to the `countLength(DNAlist, length)` but rather than counting the number of strings of the given length, this new function should return a list of the strings of that length. Here's an example of the function in action:\n",
    "\n",
    "        getLength([\"ATA\", \"ATCG\", \"TTT\", \"A\"], 3)\n",
    "        return: ['ATA', 'TTT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7839ef31-420f-4b37-92b8-b1abd32a1ff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is the coding answer to Question 3. \n",
    "def countLength(DNAlist,nuc_length=1):\n",
    "    counter=0\n",
    "    print(str(nuc_length)+\" is the length of the nt string we are seeking in this function\")\n",
    "    for nuc in DNAlist:\n",
    "        print(nuc)\n",
    "        print(counter)\n",
    "        if len(nuc)==nuc_length:\n",
    "            #print(nuc)\n",
    "            #print(counter)\n",
    "            counter=counter+1\n",
    "            print(\"loop iteration: element \"+str(counter)+ \" of the passed list is length \"+ str(nuc_length))\n",
    "        else: \n",
    "            print(\"messed up -try again\")\n",
    "    return counter  \n",
    "\n",
    "print(countLength([\"ATA\", \"ATCG\", \"TTT\", \"A\"], 3))\n",
    "print(countLength([\"ATA\", \"ATCG\", \"TTT\", \"A\"]))\n",
    "print(countLength([\"ATA\", \"ATCG\", \"TTT\", \"A\",\"ATGCGCTA\",\"TT\"]),2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd32b415-9517-475c-9008-aa6575c26b28",
   "metadata": {},
   "source": [
    "# Numpy: Extra Material\n",
    "---\n",
    "## Demo of NumPy Efficiency:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d749be5e-ab9c-4fac-839f-ca6708af450a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# Here is an example that illustrates numpy efficiency taken from this website: \n",
    "# https://towardsdatascience.com/one-simple-trick-for-speeding-up-your-python-code-with-numpy-1afc846db418\n",
    "# ------------------------------------------\n",
    "# in the first cell is using regular ol' vanilla Python and the next cell is using numpy\n",
    "# to do the same thing\n",
    "\n",
    "num_multiplies = 50000000\n",
    "data = range(num_multiplies)\n",
    "\n",
    "number = 1\n",
    "for i in data:\n",
    "    number *= 1.0000001\n",
    "\n",
    "print(number)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "596ad61f-1c94-4112-874b-7c4a60a50ac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# let's use Numpy now to do the same thing. We are bringing the module in with an alias to make it easier for ourselves\n",
    "# import numpy as alias np to save us from unnecessary typing\n",
    "import numpy as np\n",
    "\n",
    "num_multiplies = 50000000\n",
    "\n",
    "number = 1\n",
    "#numpy has array methods, but also it has general/universal methods, such as power\n",
    "number *= np.power(1.0000001, num_multiplies)\n",
    "\n",
    "print(number)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87db352a-ce14-4237-b8ce-a9ce99b9b735",
   "metadata": {},
   "source": [
    "## Examples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b4848a-3148-452c-b1cf-de58b0b83463",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --------------------------------------\n",
    "# Boring BASICS WITH NUMPY\n",
    "# --------------------------------------\n",
    "import numpy as np\n",
    "\n",
    "print(\"We can create arrays by using the array method of the NumPy: \")\n",
    "array_1 = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])\n",
    "# -------\n",
    "# 1 2 3\n",
    "# 4 5 6\n",
    "# 7 8 9 \n",
    "# -------\n",
    "# call it. \n",
    "print(array_1)\n",
    "print(\"-----Investigate some basic methods of arrays in numpy ------\")\n",
    "# multiplication over the entire array by a scalar: \n",
    "print(\"multiply array by 10\")\n",
    "print(array_1*10)\n",
    "print(\"divide array by 10\")\n",
    "print(array_1/10)\n",
    "print(\"add array_1 to array_1\")\n",
    "print(array_1+array_1)\n",
    "# element-wise since the two arrays are the same size\n",
    "print(\"multiply array_1* array_1\")\n",
    "print(array_1*array_1)\n",
    "print(\"powers of: array_1**3\")\n",
    "print(array_1**3)\n",
    "print(\"array_1 hasn't changed: \",array_1)\n",
    "#remind ourselves how many items this array has? \n",
    "# It should be 3x3 and will give us 2 dimensions (x and y)\n",
    "print(\"shape method returns a tuple of the size of each dimension\")\n",
    "print(array_1.shape)\n",
    "print(\"data type?\")\n",
    "print(array_1.dtype)\n",
    "print(\"how many dimensions?\")\n",
    "print(array_1.ndim)\n",
    "print(\"What is the size?\")\n",
    "print(array_1.size)\n",
    "#arrayname.attributename\n",
    "print(\"-----------\")\n",
    "print(\"Because they are used often, there are built in methods that create \\n a matrix of a given size of 0's and 1's\")\n",
    "array_zero = np.zeros((3,6))\n",
    "print(array_zero)\n",
    "# a vector of 1's\n",
    "array_ones = np.ones(10)\n",
    "print(array_ones)\n",
    "# We can combine these arrays:\n",
    "print(\"combine arrays:\")\n",
    "x = np.arange(0, 10) \n",
    "A = np.array([x, array_ones])\n",
    "print(A)\n",
    "# Lists prepared us for this type of data set. We can do all the things that we needed to do with lists: \n",
    "# Index: what will this return? \n",
    "print(\"we can access a particular element by using a comma-separated list \\n but I think this is a bit logically inconsistent\")\n",
    "index_1=array_1[2,2]\n",
    "print(index_1)\n",
    "#print(\"this is the more logically consistent way of accessing elements \\n form a two dimensional array\")\n",
    "# I think that the above can be confusing. The more consistent way - but slightly longer - to do it is like so: \n",
    "#index_2 = array_1[2][2]\n",
    "#print(index_2)\n",
    "print(\"We can slice by taking all rows but only the second column\")\n",
    "index_3 = array_1[:,1]\n",
    "print(index_3)\n",
    "# ---------------------- BOOLEAN LOGIC IMPOSED ON NARRAYS! --------------------------------\n",
    "print(\" We can also deploy Boolean logic on our narrays. In fact, we'll \\n look at a but more complex example in one of the following cells: \")\n",
    "# We can also use Boolean logic on narrays: \n",
    "array_bool_test = np.ones((3,3))\n",
    "print(array_bool_test>array_1)\n",
    "print(\"******\")\n",
    "print(array_1)\n",
    "print(\"Here is what happens when we impose the condition of even number on the above array\")\n",
    "print(array_1[array_1%2 ==0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea9d76ce-84bc-4f38-8d23-01f9a63f0450",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Slicing works differently in Numpy **Slicing is a done a bit differently in NumPy than in regular ol' Python: \n",
    "# if you modify an array, it is not just the copy it is actually changing the original dataset.** \n",
    "# This is mostly due to the fact that NumPy is specifically designed to be more efficient with large data sets\n",
    "# and it is *inefficient* to first copy an array and then modify the copy.  \n",
    "# slice: What will these do?\n",
    "array_1 = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])\n",
    "print('We can slice up narrays easily! Let\\'s remember what array_1 looked like')\n",
    "print(array_1)\n",
    "print(\"let's start to slice up the array above: \")\n",
    "slice_1 = array_1[:,0]\n",
    "print(\"Here is array_1[:,0]: \",slice_1)\n",
    "slice_2 = array_1[0,:]\n",
    "print(\"Here is array_1[0,:]: \",slice_2)\n",
    "slice_3 = array_1[1:3,1:3]\n",
    "print(\"Here is array_1[1:3,1:3]: \",slice_3)\n",
    "print(\"create a new narray: \")\n",
    "arr = np.arange(10)\n",
    "print(arr)\n",
    "arr[5]\n",
    "print(\"Here is arr[5]: \",arr[5])\n",
    "arr[5:8]=12\n",
    "print(arr)\n",
    "print(\" this is which is a lot like range() -  in NumPy:\")\n",
    "print(np.arange(10))\n",
    "print(np.arange(5,10))\n",
    "print(np.arange(1,10,2))\n",
    "print(\"To recreate the original array: \")\n",
    "array_2 = np.arange(1,10).reshape(3,3)\n",
    "print(array_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a98176a0-fb65-430e-aba0-cbfc318413aa",
   "metadata": {},
   "source": [
    "# Pandas: Extra Examples\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17881edf-31f7-468c-b359-23a3da258a8a",
   "metadata": {},
   "source": [
    "## Malaria Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d6af95-9a0e-4c93-9212-cfd1650b11fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here we import the malaria maize data set\n",
    "#---------------------\n",
    "# Note: your filepath will possibly look different depending on where you placed the \n",
    "# malariaMaize.csv file or if you are using colab and uploading files. \n",
    "#---------------------\n",
    "mal_maize = pd.read_csv(\"malariaMaize.csv\")\n",
    "\n",
    "# if you do not have the file saved locally, you can use the following line instead\n",
    "# mal_maize = pd.read_csv(\"https://raw.githubusercontent.com/awnorowski/BDSiC_2025/refs/heads/main/data/malariaMaize.csv\")\n",
    "\n",
    "mal_maize.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61cd0d89-4336-4ce1-b82b-1eccd88b865c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we could override the default index and insist that the index be the Villages column. \n",
    "mal_maize.set_index(\"Villages\",inplace=True)\n",
    "# Print out dataset\n",
    "print(mal_maize.head(2))\n",
    "print(\"------------------\\n\")\n",
    "print(mal_maize.tail())\n",
    "print(\"Total elements: \", mal_maize.size, \" \\nShape of the table: \", mal_maize.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c4a1890-b879-480b-b669-ff583a19ea1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# selecting\n",
    "# Print out number of cases column as Pandas Series: SINGLE SQUARE BRACKET\n",
    "print(\"Here it is as a series:\")\n",
    "print(mal_maize['Number_of_malaria_cases'])\n",
    "print(\"------------------------------\")\n",
    "\n",
    "print(\"Here it is as a DataFrame\")\n",
    "# Print out number of cases column as Pandas DataFrame: DOUBLE SQUARE BRACKET\n",
    "print(mal_maize[['Number_of_malaria_cases']])\n",
    "print(\"------------------------------\")\n",
    "# see the difference between Series and Dataframe? \n",
    "# DF includes column names, for one thing. \n",
    "\n",
    "# Print out DataFrame with number of cases and Maize_yield columns\n",
    "print(mal_maize[['Number_of_malaria_cases', 'Maize_yield']])\n",
    "print(\"malaria cases with conditions. How about only including when case > 500\")\n",
    "\n",
    "# filter: boolean filter!\n",
    "print(mal_maize[mal_maize['Number_of_malaria_cases']>500])\n",
    "# print(mal_maize.loc[mal_maize['Number_of_malaria_cases']>500])\n",
    "print(\"~~~~~~~~~~~~~\")\n",
    "# select:\n",
    "# Print out first 4 observations/rows: SLICING\n",
    "print(mal_maize[0:4])\n",
    "\n",
    "# Print out fifth, sixth, and seventh observation/row\n",
    "print(mal_maize[4:7])\n",
    "\n",
    "print(\"------------------------------\")\n",
    "\n",
    "print(\"Here is an example of iloc - although just pulling out columns is easier\")\n",
    "# Print out 3rd observation/row using index (indexing starts at 0 so this should be\n",
    "# for Wadra) \n",
    "print(mal_maize.iloc[:,0:2])\n",
    "print(\"~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\")\n",
    "# there is a difference between iloc[index position] and loc[name of position] but in this case, since the \n",
    "# index is default, they should be accessed the same way. \n",
    "print(mal_maize.loc[0:3])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda269ae-ed86-4b5f-956b-96fbca69e290",
   "metadata": {},
   "source": [
    "## Are body mass and brain mass related in mammals?\n",
    "Read in the [mammals.csv file](https://raw.githubusercontent.com/zief0002/bespectacled-antelope/main/data/mammals.csv) to investigate this question with plots. \n",
    "\n",
    "Graph body mas versus brain mass. Label the axes. We will use pandas [DataFrame.plot](https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.DataFrame.plot.html) and Mathplotlib.pyplot. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edc93d17-193f-4d56-bc34-32911c818500",
   "metadata": {},
   "outputs": [],
   "source": [
    "# in order to also have your online datasets available: \n",
    "url = \"https://raw.githubusercontent.com/zief0002/bespectacled-antelope/main/data/mammals.csv\"\n",
    "\n",
    "mammals= pd.read_csv(url)\n",
    "mammals.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcef2794-2907-47dd-9ac9-de347a030dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the plot function can take a number of different kinds: bar, hist, box,area, scatter, pie etc.  \n",
    "# once again: if you have any familiarity with R, this format will be comfortable for you!\n",
    "mammals.plot(kind = 'scatter', x = \"body_weight\", y = \"brain_weight\",title= \"Brain weight versus body weight\",color=\"orange\")\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
